### PR TITLE
Alpha version of native app generation for Mac & Windows

### PR DESCRIPTION
**Pull request by anantn**
Squashes all commits from the feature/native-app branch into one giant commits. Lots of changes, but at a high-level it enables the creating of a native .app/.exe from a XULRunner app stub included in the add-on. The resulting native app has the navigator.mozApps.amInstalled/verifyReceipt functions injected.

### CODE REVIEW

File: addons/jetpack/data/widget.html [hunk 1]
```diff
@@ -5,7 +5,7 @@
 </head>
 
 <body>
-  <div style="width: 100%; font-family: arial; font-size: 14; cursor: pointer">
+  <div style="width: 100%; font-family: arial; font-weight:bold; font-size: small; cursor: pointer">
```
Comment (mixedpuppy) [inline addons/jetpack/data/widget.html:8]: please add a bug to re-review and move style attributes into css files

File: addons/jetpack/lib/api.js [hunk 2]
```diff
@@ -41,6 +41,10 @@ Cu.import("resource://gre/modules/XPCOMUtils.jsm");
 
 var { TypedStorage } = require("typed_storage");
 
+//don't remove this please!
```
Comment (mixedpuppy) [inline addons/jetpack/lib/api.js]: why?

File: addons/jetpack/lib/api.js [hunk 3]
```diff
@@ -51,6 +55,7 @@ if (!console || !console.log) {
 
 var { Manifest } = require("./manifest");
 var { URLParse } = require("./urlmatch");
+var { NativeShell } = require("./nativeshell");
```
Comment (mixedpuppy) [inline addons/jetpack/lib/api.js:52]: why twice?

File: addons/jetpack/lib/injector.js [hunk 4]
```diff
@@ -41,12 +41,22 @@
 /* Inject the People content API into window.navigator objects. */
 /* Partly based on code in the Geode extension. */
 
-const { Cc, Ci, Cu } = require("chrome");
-Cu.import("resource://gre/modules/XPCOMUtils.jsm");
-const xulApp = require("api-utils/xul-app");
+var HAS_NAVIGATOR_INJECTOR;
+if (typeof require !== "undefined") {
```
Comment (mixedpuppy) [inline addons/jetpack/lib/injector.js:45]: any reason to have this code do the same thing two different ways?  We should just use the pure-xpcom version and reduce the code here.

File: addons/jetpack/lib/main.js [hunk 5]
```diff
@@ -493,5 +493,8 @@ function shutdown(why) {
 // Let's go!
 startup(addon.data.url);
 
+const socketserver = require("socketserver");
+socketserver.startServer();
+
```
Comment (mixedpuppy) [inline addons/jetpack/lib/main.js]: what is this socket server for, and why is it being started at the top level?  Is it necessary to start at the top level?

File: addons/jetpack/lib/nativeshell.js [hunk 6]
```diff
@@ -0,0 +1,902 @@
...
+  } catch(e) {
+    throw("writeFile - "
+        + "Failed writing file to "
+        + destFile
+        + " (" + e + ")");
+  }
+}
+
+//ASYNC file reading/writing/copying code.  unable to evaluate the issues that might occur during a file
+// tree copy, so putting on hold for now.
```
Comment (mixedpuppy) [inline addons/jetpack/lib/nativeshell.js]: lets please not propagate dead code

File: addons/jetpack/lib/socketserver.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
+const {Cc, Ci} = require("chrome");
+var pageWorkers = require("page-worker");
+
+var gIdServer;
+const PORT_RANGE_START = 7350;
+const PORT_RANGE_END = 7400;
+
+function IdentityServer() {
```
Comment (mixedpuppy) [inline addons/jetpack/lib/socketserver.js:48]: some documentation on what this identityserver is doing would be useful.

File: Makefile [hunk 8]
```diff
@@ -44,6 +44,7 @@ endif
 all: xpi
 
 xpi:    pull
+	${MAKE} -C addons/jetpack/mac/
```
Comment (mixedpuppy) [inline Makefile]: does this build on all platforms?

File: addons/jetpack/data/native-install/XUL/content/main.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
+
+var Cu = Components.utils;
+var Cc = Components.classes;
+var Ci = Components.interfaces;
+
+var os = Cc["@mozilla.org/xre/app-info;1"].getService(Ci.nsIXULRuntime).OS;
+    
+// only make the tiny window that handles window closing events (and quits the app) for Mac OS
```
Comment (mixedpuppy) [inline addons/jetpack/data/native-install/XUL/content/main.js:8]: The last I worked on xulrunner stuff, there was a hidden window for osx that you could use for things like this, worth looking into further

File: addons/jetpack/data/native-install/XUL/content/window.js [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+      } catch (e) {
+          alert(e);
+          port++;
+          attemptConnection();
+      }
+  }
+  attemptConnection();
+}
+
+/*
```
Comment (mixedpuppy) [inline addons/jetpack/data/native-install/XUL/content/window.js]: dead code?

File: addons/jetpack/lib/main.js (hunk 5 above)
Comment (ianb) [inline addons/jetpack/lib/main.js]: @michaelrhanson can probably answer this.

Comment (mixedpuppy) [inline addons/jetpack/lib/main.js]: ok, per discussion in irc, lets move socketserver startup into startup().  We also need an unloader, and it seems stopServer is not complete.  lets get bugs on that issue.

File: addons/jetpack/lib/socketserver.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+		}
+	}
+}
+
+function startServer() {
+	gIdServer = new IdentityServer();
+	gIdServer.start();
+}
+function stopServer() {
+
```
Comment (mixedpuppy) [inline addons/jetpack/lib/socketserver.js:174]: a bug for implementing stopServer, and make sure we add an unlaoder in startup()

**Comment by mixedpuppy** [comment]
Overall comments.  Dead code needs to go.  large source code files with no documentation are PITA.  Comments without explanation are useless (e.g. "Dont remove this please!").  With this code going into production I see code-debt being built.  I'm not going to hold up the merge, but I hope some of this gets addressed.

File: addons/jetpack/lib/api.js (hunk 2 above)
Comment (anantn) [inline addons/jetpack/lib/api.js]: Fixed

File: addons/jetpack/lib/api.js (hunk 3 above)
Comment (anantn) [inline addons/jetpack/lib/api.js:52]: Fixed

File: addons/jetpack/lib/injector.js (hunk 4 above)
Comment (anantn) [inline addons/jetpack/lib/injector.js:45]: this is possibly related to the injection not working in native apps, the pure-xpcom version seems to be broken but the jetpack version works. will look into fixing this and using a xpcom-only file for both the add-on & native apps.

File: addons/jetpack/lib/nativeshell.js (hunk 6 above)
Comment (anantn) [inline addons/jetpack/lib/nativeshell.js]: Removed

File: addons/jetpack/lib/socketserver.js (hunk 7 above)
Comment (anantn) [inline addons/jetpack/lib/socketserver.js:48]: Added a small note at the top of the file, will add more docs later.

File: Makefile (hunk 8 above)
Comment (anantn) [inline Makefile]: Nope! Fixed.

File: addons/jetpack/data/native-install/XUL/content/window.js (hunk 10 above)
Comment (anantn) [inline addons/jetpack/data/native-install/XUL/content/window.js]: Removed. Was mostly a reminder to implement navigator.id.\* for native apps, will file a bug for that.

**Comment by anantn** [comment]
Cool, thanks! WIll make bugs for issues that were not addressed in the last commit.

File: addons/jetpack/lib/socketserver.js [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+		var port = PORT_RANGE_START;
+		while (port < PORT_RANGE_END)
+		{
+			try {
+				this.serverSocket.init(port, true, 5);
+				console.log("Started identity server on port " + port);
+				break;
+			} catch (e) {
+				port++;
+				if (port == PORT_RANGE_END)
```
Comment (kamathln) [inline addons/jetpack/lib/socketserver.js:68]: Did you get the if and else mixed up?

**Pull request merged**